shell/src/main/java/org/apache/accumulo/shell/commands/FateCommand.java

```diff
@@ -136,7 +140,29 @@ public int execute(final String fullCommand, final CommandLine cl, final Shell s
         getZooReaderWriter(context, siteConfig, cl.getOptionValue(secretOption.getOpt()));
     ZooStore<FateCommand> zs = new ZooStore<>(path, zk);
 
-    if ("fail".equals(cmd)) {
+    if ("fail-live".equals(cmd)) {
```

Comment:
Why not just call it cancel?

Reply:
I'm not tied to the name, we can change it to whatever... Doesn't the currently named `fail` method actually cancel ? I was trying to stick with the current naming convention.

Reply:
I do not know if there is a nuance with cancel vs. fail.  Cancel seems (to me) a more normal operation - I changed my mind. Fail seems more of an error condition.  The documentation in the FATE section states that  "Failing an operation is not a normal procedure" Does canceling a compaction raise to that level?  So either way fail / or cancel, but it maybe worth considering the context and usage.   

One one hand, not needing to remember when its fail vs. cancel (follow existing naming) is a positive - unless cancel is more "normal" and does not take a lot of insight for an operator to execute.

Reply:
SInce I equated `fail` with `cancel` I tried to use a name that implied the user could cancel/fail the operation with the manager up. Hence `fail-live`. That sounded better than `fail-up`, lol.

Reply:
To Ed's question, "Does canceling a compaction raise to that level?" ... I think both operations are pretty abnormal... to the point that I'm still wondering whether we really need to be able to do this live when taking the manager offline is probably fine, and safest, under abnormal conditions. I worry that people will normalize it, and their systems will become unstable over time by abusing this.

Reply:
So, we give users the ability to create fate transactions easily. I think in only one case, we give them the ability to cancel them (cancel compactions). Taking the manager down to fail a fate transaction is very disruptive (and doesn't it terminate the FaTE transaction runners in the middle of their operations?). Certainly the changes I made in #2462 will let users identify which transactions have never been started, those should be able to be cancelled with no issue.

Reply:
@dlmarion wrote:
> So, we give users the ability to create fate transactions easily.

I don't think that's an accurate characterization. Users don't create fate transactions at all. The Manager does so to service some user requests, like "create a table". This distinction matters, because fate transactions are an implementation detail and should largely be hidden from the user. In many cases, these operations are expected to be so fast (like creating a table), the user shouldn't even have the concept of it being a queued action. The only reason we run them as a fate operation is so we don't leave Accumulo in a dirty state in the case of a failure. In the create table example, users don't need to be able to cancel the fate operations, because they can just delete an unwanted table.

So, in a lot of cases, the fate operations aren't actually there for the user. They are there for Accumulo's internal needs. Implementation detail. So, it makes a kind of sense that users don't directly cancel them, or can't, under normal circumstances.

>  I think in only one case, we give them the ability to cancel them (cancel compactions).

The reason users can easily cancel compaction is because the semantics for the compaction API is a "queue this compaction" request, not an "execute this compaction immediately" request. The cancellation for compactions is available to users because of the semantics of the compaction API, not because of the implementation detail of it using fate to execute it. Those queuing and cancelling semantics aren't generalizable to all actions that we perform via fate and should be considered on a case-by-case basis and exposed through our normal APIs, just like compactions are today. Interacting with Fate is not going through normal APIs. It's surgery, and should be rare.

> Taking the manager down to fail a fate transaction is very disruptive (and doesn't it terminate the FaTE transaction runners in the middle of their operations?).

Aborting transaction runners in the middle of an operation is generally safe because FaTE RepOs are designed to be idempotent and replayable to resume a transaction. They also have the ability to roll back when they are move into a failed state. The reason we kill the manager is to ensure we know the current state of any transactions that are in progress that we intend to fail, so there aren't any race conditions or risks of some side-effect of a poorly behaved interrupted thread. That's generally acceptable because aborting fate operations is supposed to be rare, and the surgery should be relatively safe, with predictable outcomes. But, it's still surgery, and we don't exactly want to lower the barrier to entry to surgery.

Fate is not intended to be a general request queue that users add and remove stuff from at will, they are for internal consistency. When we fail stuff, it's because we're trying to force something that isn't normal. If we want some operations to be cancel-able, I think maybe creating a dedicated asynchronous API for the operations we want to be able to cancel might be a better approach, than trying to lower the barrier to entry to Fate surgery or trying to generalize Fate operations like it's a general purpose executor service.

> Certainly the changes I made in #2462 will let users identify which transactions have never been started, those should be able to be cancelled with no issue.

Is the intent of this current PR to only be able to `fail-live` things that have never been started, or operations in any state?  If it's only to help clear out unstarted things, I think I'd be more enthusiastic about it. It might even be useful have a safe live option for `cancel all transactions not yet started`.

Reply:
Generally the two operations that would be useful to cancel would be compactions and bulk-import - both which can be long running operations, that after they have been submitted, things could have changed such that they are not longer desired - or at least not desired at that point in time.

It would also be desirable that a delete operation take precedence over any other operation in progress or queued for a table.  If a compaction or a bulk import is running and then the system requests the table be deleted, the only consistency required is that the metadata table remains correct and the proper files get marked for gc'ing - anything else that is occurring is work that is going to be thrown away.  There is does not much benefit of having the compaction continue to read / write data, that as soon as it completes is going to be marked for deletion.  There may be subtleties that I'm ignoring if the table was cloned, but in practice, cloning seems to be rare,

Reply:
> Generally the two operations that would be useful to cancel would be compactions and bulk-import

So, we currently have the ability to cancel compactions. I don't believe the new bulk import is long-running. So, any problems with long-running bulk imports should be alleviated by avoiding the old bulk import mechanism.

> It would also be desirable that a delete operation take precedence over any other operation in progress or queued for a table.

Agreed, but I'm not sure that would require exposing anything to users/admins. That should just be handled internally. I think recent work has moved in this direction.

> There may be subtleties that I'm ignoring if the table was cloned

Cloning should be fast, but it may wait on other operations, like compactions, just to avoid having to do its own redundant compactions. I'm not sure. I'd have to check the code, but a sensible API for that might be to add an option to wait for outstanding compactions (or to not wait).

Reply:
Feedback from user(s) is:
  1. cancelling a compaction from the compact command has never worked.
    > Is it broken in 1.x versions?
  2. Stopping the Manager is a problem because they can't load data via bulk imports, they have strict timeliness requirements. They have to stop bulk imports from getting kicked off, wait for the current ones to finish, then take down the Manager. Stopping the Manager also stops any long-running WAL recovery due to a failed TServer, table balancing, etc.
  3. Another case is issuing DeleteRows command to remove  some data that is causing problems system-wide. The DeleteRows operation may be blocked by less important Fate transactions which may be operating on a different range within the table.
  4. Accumulo's approach to locking the entire table seems heavy-handed.  There could be bulk imports and compactions running that affect one active part of a table (let's say a newer date) and we're trying to deleterows or merge on a completely different range that's inactive (older dates).
    > Maybe we should support Row or Range locks?

Reply:
Adding ranges to the tables locks that FATE acquires seems like a good thing to explore.

Reply:
> Feedback from user(s) is:
> 
> 2. Stopping the Manager is a problem because they can't load data via bulk imports, they have strict timeliness requirements. They have to stop bulk imports from getting kicked off, wait for the current ones to finish, then take down the Manager. Stopping the Manager also stops any long-running WAL recovery due to a failed TServer, table balancing, etc.

I understand stopping the manager is not preferred. I'm not sure lowering the bar to surgery by allowing it to be performed on a conscious patient is the right solution, though (might be taking the metaphor too far here :smiley_cat: ). I'd rather focus efforts on surgery-prevention, if we can. For example, using the new bulk imports, using the new working compaction queue cancellation, figuring out why 1.x compactions can't be cancelled, etc., would be a great start. Adding range locks for compaction/merge/delete would also be great.

Right now, this seems very risky to add, especially since we haven't even ironed out the kinks in the other recent changes in these areas. ExternalCompactions ITs are still flaky, and now FateIT is failing too. These imply that there is instability in this code. Giving users more direct access to making live changes to internals, when we haven't ironed out those internals in our own tests, seems very high risk, and the wrong solution to the underlying problems users are seeing. If all those other things were working well, I would be less opposed to taking the risk here, but they're not. So, this risk just seems too much right now.

Reply:
> might be taking the metaphor too far here

I'm not convinced that it's the right one as it's [documented](https://accumulo.apache.org/docs/2.x/administration/fate). For me, surgery is manually inserting / updating / deleting rows from the metadata table.

> Right now, this seems very risky to add, especially since we haven't even ironed out the kinks in the other recent changes in these areas. ExternalCompactions ITs are still flaky, and now FateIT is failing too.

I"m starting to think the issue is ZK based on the information you added to in #2463 and the fact that FateIT has one test.

Reply:
I think a path forward might be to:

1. Verify that compaction cancellations are working in 1.x and 2.1.0 (new GH issue)
2. Modify this PR to only cancel FaTE transactions in the submitted state
3. Research range lock implementation  (new GH issue)

Reply:
This PR has been modified (2) and I have submitted issues #2482 (1) and #2483 (3)